### ISSUE TITLE
Control anchor points should not be limited from 0 to 1

### ISSUE THREAD

**Issue by mr746866** [godotengine/godot#21080]
I'll try to explain the issue using an example.

Suppose, you have a parent `Control` and a child `Control`. The parent `Control` will change its `rect_size` dynamically, both width and height. The child `Control` needs to always satisfy the following conditions.

1. The child width must always be 40% of parent width
2. The child height must always be 40% of parent height
3. The child top-left corner must always be at 20% of parent width and 20% of parent height

This conditions are easily met by setting anchor points and margins on the child `Control`.

![screen shot 2018-08-16 at 5 33 04 pm](https://user-images.githubusercontent.com/11366464/44206601-3579ee00-a17c-11e8-89e3-c567471f93d2.png)
![screen shot 2018-08-16 at 5 27 04 pm](https://user-images.githubusercontent.com/11366464/44206609-3b6fcf00-a17c-11e8-8eab-1df27e053603.png)
![screen shot 2018-08-16 at 5 27 10 pm](https://user-images.githubusercontent.com/11366464/44206618-3f035600-a17c-11e8-96b5-a175dfed99e3.png)
![screen shot 2018-08-16 at 5 27 17 pm](https://user-images.githubusercontent.com/11366464/44206624-4296dd00-a17c-11e8-94a6-7dbb0d204a8c.png)

Very easy solution, right? This method can be used as long as the child needs to be inside the parent rect. But the situation changes completely when part of the child needs to be outside of parent rect.

For example, just change the conditions to the following.

1. The child width must always be 40% of parent width
2. The child height must always be 40% of parent height
3. The child top-left corner must always be at 80% of parent width and 80% of parent height

If Godot didn't limit the anchor points from 0 to 1, it would've been sufficient to just change the anchor points to the following.

![screen shot 2018-08-16 at 5 33 04 pm copy](https://user-images.githubusercontent.com/11366464/44207989-ea160e80-a180-11e8-87dd-4920c087da84.png)

But as it does not allow anchor points to be outside of this range, we need to find alternative solutions like using scripts or wrapping the child and/or parent `Control`s inside `Node2D`.

Similarly, if we needed the child rect to be outside the left and top border of the parent rect, we could use negative anchor points if Godot allowed it.

**TLDR:** Removing the limit on anchor points will make creating percentage based UI layouts where the child rect has parts outside parent rect a lot easier. With the limit in place, its unnecessarily complex.

**Comment by groud** [comment]
I have no clue what your use case is. 

Having to set the size of a child control which is outside of its parent seems a very corner case to me. And it is probably easily workaroundable easily by changing the control node hierarchy (moving the parent node at the same level as its child and adding another parent node that encompasses all of them) Consequently, I see no point in making the common case unnecessarily complex for something that is easily worked around via script or changing the nodes hierarchy.

**Comment by groud** [comment]
Well, it requires the anchors to be able to move outside the parent, which complexifie the workflow when you simply want a control to stick to a side of its parent. And stick a control to a side is a lot more common than having children outside of their parent. 
It also requires changing the border values in the inspector (for the sliders), for arbitrary values. 

But still, I don't get the limitation of the current system, which is simpler than what you propose.

**Comment by mr746866** [comment]
@groud I think you misunderstood me, I am not suggesting to totally change the anchor points system. Whatever you can do in current system, you can do as well in my suggested system, and without any change in value. If you want a control to stick to a side of its parent, you can do so in my suggested system, in exactly the same way as you do in current system.

I just want to be able to set anchor points negative, as well as over 1. This is not a new idea by the way, and a lot of other game engines allow any floating number (both negative and positive) as anchor points.

**Comment by groud** [comment]
I did no misunderstood you. Allowing values outside 0-1 is a bad idea to me. Clamping the anchors to 0-1 is a lot simpler when you drag the anchors to a side, than having to miticulously target for the 0 or 1 values. Because it is a really common case, and because your initial problem is easily workaroundable, it is not, IMHO, worth allowing values outside 0-1.

**Comment by groud** [comment]
To be more precise, I'd be OK with that change only if it does not modifies the current way of editing the properties (the 2d editors anchors are clamped and the sliders go from 0 to 1). But you may be able to enter custom values by hand, a little bit like how it's done in blender.

**Comment by groud** [comment]
According to @akien-mga this is possible, so I'll implement that.

**Comment by mr746866** [comment]
@groud Thanks, hope its an easy task.

**Comment by mr746866** [comment]
@groud I want to make a PR for this issue but want to discuss 2 things first.

1. Should the anchor values be limited within a range? Like the margin values are limited from -4096 to 4096?

2. Should the 2D visual editor's anchors be clamped within parent rect? Or is it enough to just snap?

**Comment by Chaosus** [comment]
Ok, I write my opinion
1. In editor it make sense, in gdscript no - its breaks compatibility
2. If its improve usability then its good
So, feel free to create PR :)

**Comment by mr746866** [comment]
@Chaosus 

1. What should be the limit in editor? Is -4096.0 to 4096.0 good?

2. I think just snapping the anchors is best here, no need to clamp.

**Comment by Chaosus** [comment]
Better ask @groud about something specific ^^

**Comment by groud** [comment]
1) no, the limits should be 0-1 in the editor. You can allow higher or lower typed values with the "or_greater" and "or_less" keywords in the editor hints. 

2) The anchors should stay clamped in the editor

I don't want this feature to impact the workflow. So only manually typed values are allowed to go beyond 0 or 1

**Comment by mr746866** [comment]
Suppose the user sets `anchor_right` to 1.2 and `anchor_bottom` to 1.5 by typing. To reflect these values, in the 2D editor the anchor pins are now outside the parent rect.

Now, what happens if the user wants to change the anchors using the anchor pins?

**Comment by mr746866** [comment]
@groud Where are the "or_greater" and "or_lesser" keywords by the way? I don't see them used anywhere in source.

**Comment by groud** [comment]
> Now, what happens if the user wants to change the anchors using the anchor pins?

When they try to grab it the value is then clamped. Maybe a special key might be used to allow moving anchors outside though, but I think ctrl, alt and shift should be used for more obvious purpose such as symmetrical resize or centered resize. This could be added later on.

**Comment by ElfEars** [comment]
I'd +1 this suggestion. Sliding UI stuff on and off screen is a very common UI use case and currently you have to get hacky if you want it to work on all screens.
